src/App.tsx

```diff
@@ -1,26 +1,213 @@
+
+    setIsLoading(true);
+    setFocusHeader(false);
+
+    addTodo(newTodo)
+      .then(response => {
+        setTodos(currentTodos => [...currentTodos, response]);
+        setTodoInputValue('');
+      })
+      .catch(() => {
+        handleError(Error.ADDING_TODO);
+      })
+      .finally(() => {
+        setIsLoading(false);
+        setTempTodo(null);
+        setFocusHeader(true);
+      });
+  };
+
+  const handleDeleteTodo = (id: number) => {
+    setIsLoading(true);
+    setLoadingTodoIds(current => [...current, id]);
+    setFocusHeader(false);
+
+    deleteTodo(id)
+      .then(() => {
+        setTodos(currentTodos => currentTodos.filter(todo => todo.id !== id));
+
+        if (editingTodoId) {
+          setEditingTodoId(null);
+        }
+      })
+      .catch(() => handleError(Error.DELETING_TODO))
+      .finally(() => {
+        setLoadingTodoIds(current =>
+          current.filter(deletedId => deletedId !== id),
+        );
+        setIsLoading(false);
+        if (!editingTodoId) {
+          setFocusHeader(true);
+        }
+      });
+  };
+
+  const handleDeleteCompleted = () => {
+    const completedIds = todos.reduce<number[]>((ids, todo) => {
+      if (todo.completed) {
+        ids.push(todo.id);
+      }
+
+      return ids;
+    }, []);
+
+    completedIds.forEach(handleDeleteTodo);
+  };
+
+  const handleUpdateTodo = (id: number, data: Partial<Todo>) => {
+    setIsLoading(true);
+    setLoadingTodoIds(currentIds => [...currentIds, id]);
+
+    updateTodo({ id, ...data })
+      .then(updatedTodo => {
+        setTodos(currentTodos =>
+          currentTodos.map(todo =>
+            todo.id === updatedTodo.id ? updatedTodo : todo,
+          ),
+        );
+
+        setEditingTodoId(null);
+      })
+      .catch(() => handleError(Error.UPDATING_TODO))
+      .finally(() => {
+        setLoadingTodoIds(ids =>
+          ids.filter(loadingTodoId => loadingTodoId !== id),
+        );
+        setIsLoading(false);
+      });
+  };
+
+  const handleToggleStatus = (id: number, currentStatus: boolean) => {
+    handleUpdateTodo(id, { completed: !currentStatus });
+  };
+
+  const handleToggleAll = () => {
+    const isAllStatusSame =
+      todosAmount === uncompletedTodosAmount || !uncompletedTodosAmount;
+
+    todos.forEach(({ id, completed }) => {
+      const newStatus = isAllStatusSame ? !completed : true;
+
+      if (completed !== newStatus) {
+        handleUpdateTodo(id, { completed: newStatus });
+      }
+    });
+  };
+
+  const handleUpdateTitle = (todo: Todo, title: string) => {
+    if (!title.length) {
+      handleDeleteTodo(todo.id);
+
+      return;
+    }
+
+    if (todo.title === title) {
+      setEditingTodoId(null);
+
+      return;
+    }
+
+    handleUpdateTodo(todo.id, { title: title });
+  };
+
+  useEffect(() => {
+    getTodos()
+      .then(currentTodos => {
+        setTodos(currentTodos);
+      })
+      .catch(() => {
+        handleError(Error.LOADING_TODOS);
+      });
```

Comment:
It's better to move the logic inside the `useEffect `hook to a separate function, as it improves code readability, reusability, and makes testing easier. It also helps keep `useEffect `focused on side effects.